File: src/linkedlist/Solution142.java
package linkedlist;

import java.util.HashSet;

// 142. Linked List Cycle II
public class Solution142 {

    public ListNode detectCycle(ListNode head) {

        HashSet<ListNode> set = new HashSet<>();
        ListNode fast = head;
        ListNode slow = head;

        while (fast != null && fast.next != null) {
            set.add(slow);
            slow = slow.next;
            if (set.contains(fast.next)) return fast.next;
            if (set.contains(fast.next.next)) return fast.next.next;
            fast = fast.next.next;
        }

        return null;
    }
}

File: python/linked_list/Solution142.py
# 142. 环形链表 II
from typing import Optional

from linked_list import ListNode


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        detect = set()
        fast = head
        slow = head
        while fast and fast.next:
            detect.add(slow)
            slow = slow.next
            if fast.next in detect:
                return fast.next
            if fast.next.next in detect:
                return fast.next.next
            fast = fast.next.next
        return None
